BUG: `sftp_write` fails if the destination file does not exist in the container

## Problem statement
 `sftp_write`  method that we are using in robottelo behaves differently based on the broker provider - Container or AAP.
 
https://github.com/SatelliteQE/robottelo/blob/eca242110a5c9b4040d62d04a7094d00cebf5aa7/robottelo/hosts.py#L686-L697

### Traceback
```
../../lib64/python3.8/site-packages/docker/api/client.py:268: in _raise_for_status
    response.raise_for_status()
../../lib64/python3.8/site-packages/requests/models.py:1021: in raise_for_status
    raise HTTPError(http_error_msg, response=self)
E   requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http+docker://ssh/v1.41/containers/9e2407883fed4af87a97ab80579150153a3ef2a7fbdcd0558618ac6b3bb3762e/archive?path=%2Fetc%2Frhsm%2Ffacts%2Flocations.facts

The above exception was the direct cause of the following exception:
tests/foreman/ui/test_dashboard.py:257: in test_positive_user_access_with_host_filter
    repos_collection.setup_virtual_machine(
robottelo/host_helpers/repository_mixins.py:725: in setup_virtual_machine
    vm.contenthost_setup(
robottelo/hosts.py:1007: in contenthost_setup
    self.set_facts({'locations.facts': {'foreman_location': str(location_title)}})
robottelo/host_helpers/contenthost_mixins.py:145: in set_facts
    self.put(tf.name, f'/etc/rhsm/facts/{filename}')
robottelo/hosts.py:697: in put
    self.session.sftp_write(source=local_path, destination=remote_path)
../../lib64/python3.8/site-packages/broker/session.py:234: in sftp_write
    self._cont_inst._cont_inst.put_archive(str(destination), tar.read_bytes())
../../lib64/python3.8/site-packages/docker/models/containers.py:334: in put_archive
    return self.client.api.put_archive(self.id, path, data)
../../lib64/python3.8/site-packages/docker/utils/decorators.py:19: in wrapped
    return f(self, resource_id, *args, **kwargs)
../../lib64/python3.8/site-packages/docker/api/container.py:976: in put_archive
    self._raise_for_status(res)
../../lib64/python3.8/site-packages/docker/api/client.py:270: in _raise_for_status
    raise create_api_error_from_http_exception(e) from e
../../lib64/python3.8/site-packages/docker/errors.py:39: in create_api_error_from_http_exception
    raise cls(e, response=response, explanation=explanation) from e
E   docker.errors.NotFound: 404 Client Error for http+docker://ssh/v1.41/containers/9e2407883fed4af87a97ab80579150153a3ef2a7fbdcd0558618ac6b3bb3762e/archive?path=%2Fetc%2Frhsm%2Ffacts%2Flocations.facts: Not Found ("Could not find the file /etc/rhsm/facts/locations.facts in container 9e2407883fed4af87a97ab80579150153a3ef2a7fbdcd0558618ac6b3bb3762e")****
```

If the file does not exist on the host, the above exception is raised.

## Proposed solution
Broker should be smart and should create the file in the container if it does not exist - like in the VM (AnsibleTower provider).

## Notes
@peterdragun and I played around with it a bit and we found out that if the `destination` file exists, it is overridden by the `source` file.

--- a/broker/session.py
+++ b/broker/session.py
@@ -81,6 +81,8 @@ class Session:
         """read a remote file into a local destination"""
         if not destination:
             destination = source
+        elif destination.endswith("/"):
+            destination = destination + Path(source).name
         # create the destination path if it doesn't exist
         destination = Path(destination)
         destination.parent.mkdir(parents=True, exist_ok=True)
@@ -94,19 +96,25 @@ class Session:
                 for size, data in remote:
                     local.write(data)
 
-    def sftp_write(self, source, destination=None):
+    def sftp_write(self, source, destination=None, ensure_dir=True):
         """sftp write a local file to a remote destination"""
         if not destination:
             destination = source
+        elif destination.endswith("/"):
+            destination = destination + Path(source).name
         data = Path(source).read_bytes()
+        if ensure_dir:
+            self.run(f"mkdir -p {Path(destination).absolute().parent}")
         sftp = self.session.sftp_init()
         with sftp.open(destination, FILE_FLAGS, SFTP_MODE) as remote:
             remote.write(data)
 
-    def remote_copy(self, source, dest_host):
+    def remote_copy(self, source, dest_host, ensure_dir=True):
         """Copy a file from this host to another"""
         sftp_down = self.session.sftp_init()
         sftp_up = dest_host.session.session.sftp_init()
+        if ensure_dir:
+            dest_host.run(f"mkdir -p {Path(source).absolute().parent}")
         with sftp_down.open(
             source, ssh2_sftp.LIBSSH2_FXF_READ, ssh2_sftp.LIBSSH2_SFTP_S_IRUSR
         ) as download:
@@ -114,10 +122,12 @@ class Session:
                 for size, data in download:
                     upload.write(data)
 
-    def scp_write(self, source, destination=None):
+    def scp_write(self, source, destination=None, ensure_dir=True):
         """scp write a local file to a remote destination"""
         if not destination:
             destination = source
+        elif destination.endswith("/"):
+            destination = destination + Path(source).name
         fileinfo = os.stat(source)
         chan = self.session.scp_send64(
             destination,
@@ -126,6 +136,8 @@ class Session:
             fileinfo.st_mtime,
             fileinfo.st_atime,
         )
+        if ensure_dir:
+            self.run(f"mkdir -p {Path(destination).absolute().parent}")
         with open(source, "rb") as local:
             for data in local:
                 chan.write(data)
@@ -221,7 +233,7 @@ class ContainerSession:
         """Needed for simple compatability with Session"""
         pass
 
-    def sftp_write(self, source, destination=None):
+    def sftp_write(self, source, destination=None, ensure_dir=True):
         """Add one of more files to the container"""
         # ensure source is a list of Path objects
         if not isinstance(source, list):
@@ -232,15 +244,17 @@ class ContainerSession:
         for src in source:
             if not Path(src).exists():
                 raise FileNotFoundError(src)
-        destination = Path(destination) or source[0].parent
+        destination = destination or f"{source[0].parent}/"
         # Files need to be added to a tarfile
         with helpers.temporary_tar(source) as tar:
             logger.debug(
                 f"{self._cont_inst.hostname} adding file(s) {source} to {destination}"
             )
-            # if the destination is a file, create the parent path
-            if destination.is_file():
-                self.execute(f"mkdir -p {destination.parent}")
+            if ensure_dir:
+                if destination.endswith("/"):
+                    self.run(f"mkdir -m 666 -p {destination}")
+                else:
+                    self.run(f"mkdir -m 666 -p {Path(destination).parent}")
             self._cont_inst._cont_inst.put_archive(str(destination), tar.read_bytes())
 
     def sftp_read(self, source, destination=None):

--- a/tests/functional/test_containers.py
+++ b/tests/functional/test_containers.py
@@ -84,6 +84,6 @@ def test_container_e2e_mp():
             res = c_host.execute("hostname")
             assert res.stdout.strip() == c_host.hostname
             # Test that a file can be uploaded to the container
-            c_host.session.sftp_write("broker_settings.yaml", "/root")
-            res = c_host.execute("ls")
+            c_host.session.sftp_write("broker_settings.yaml", "/tmp/fake/")
+            res = c_host.execute("ls /tmp/fake")
             assert "broker_settings.yaml" in res.stdout

--- a/tests/functional/test_satlab.py
+++ b/tests/functional/test_satlab.py
@@ -68,3 +68,6 @@ def test_tower_host():
     with Broker(workflow="deploy-base-rhel") as r_host:
         res = r_host.execute("hostname")
         assert res.stdout.strip() == r_host.hostname
+        r_host.session.sftp_write("broker_settings.yaml", "/tmp/fake/")
+        res = r_host.execute("ls /tmp/fake")
+        assert "broker_settings.yaml" in res.stdout

